### PR TITLE
[ORCA] Fix boolean testing when the child is not an ident but an expression

### PR DESCRIPTION
**Pull request by fishtree1161**
After the commit "[ORCA] Support boolean static partition pruning (#15348)", we added support for boolean testing in partition pruning or other condition pruning. However, currently, we only treated it as a parent of ident. When its child is an expression, errors may occur.

For example 1, this query will lead to a "One-Time Filter: false" because we treated it as "b is true and (not b)".

create table test (a int, b boolean);
explain select * from test where (not b) is true and (not b);

For example 2, this query will have no results because partition pruning will prune the current part as we treated it as "col2 IS true".

CREATE TABLE pt_bool_tab (
  col1 int,
  col2 bool
)
distributed by (col1)
partition by list(col2)
(
  partition part1 values(true),
  partition part2 values(false)
);
INSERT INTO pt_bool_tab SELECT i, true FROM generate_series(1,3)i;
INSERT INTO pt_bool_tab SELECT i, false FROM generate_series(1,2)i;
SELECT * FROM pt_bool_tab WHERE (not col2) IS true;

To fix this issue, we can use the following approaches to replace boolean testing: "expr (with infer_nulls_as)", "not (with infer_nulls_as)", "is null", and "is not null". 
This is the table about results of boolean tests & some other boolean operators

Parameter		T	F	NULL
----------------------------
is true			T	F	F
is not false		T	F	T
is false			F	T	F
is not true		F	T	T
is unknown		F	F	T
is not unknown	T	T	F
----------------------------
expr				T	F	infer_nulls_as
not				F	T	infer_nulls_as
is null			F	F	T
is not null		T	T	F

So, we treat
(is true / is not false)	as	expr with infer_nulls_as
(is false / is not true)	as	not with infer_nulls_as
is unknown 					as	is null
is not unknown 				as	is not null

TODO: After this fix, there is also one optimization remaining. The way of "is null" cannot process pruning when its child is an expression. It is difficult to differentiate between ((ident is true) is null) and (ident is null) when ident is null. We may consider adding a null_is_null flag in CConstraintInterval in the future.

### CODE REVIEW

**Review by dgkimura — APPROVED** [review]
Nice find!  I really like the added test coverage.  And the fix looks reasonable to me.

File: src/backend/gporca/libgpopt/src/base/CConstraintInterval.cpp [hunk 1]
```diff
@@ -413,13 +413,18 @@ CConstraintInterval::PciIntervalFromScalarNullTest(CMemoryPool *mp,
 												   CColRef *colref)
 {
 	GPOS_ASSERT(nullptr != pexpr);
-	GPOS_ASSERT(CUtils::FScalarNullTest(pexpr));
+	// just let (is unknown & is not unknown) in
+	GPOS_ASSERT(CUtils::FScalarNullTest(pexpr) ||
+				(pexpr->Pop()->Eopid() == COperator::EopScalarBooleanTest &&
```
[the remark this answers is not in the document]
Comment (fishtree1161) [reply src/backend/gporca/libgpopt/src/base/CConstraintInterval.cpp]: Thanks for your review!
I changed PciIntervalFromScalarNullTest to PciIntervalFromScalarNullAndUnknownTest,
and add some comment for this method.

**Review by gpopt — APPROVED** [review]
Can you refine the git commit message to 80 characters width? Thanks for the contribution.

**Comment by fishtree1161** [comment]
> Can you refine the git commit message to 80 characters width? Thanks for the contribution.

I have formatted the code and git commit message, please review it again.

**Comment by dgkimura** [comment]
Merged. Thanks for the contribution!

**Comment by fishtree1161** [comment]
> Merged. Thanks for the contribution!

Thanks for formatting, it's my pleasure to contribute.

**Pull request merged**